src/app/utilities/moodle/helper.ts

```diff
@@ -0,0 +1,175 @@
+import {
+  MoodleCourse,
+  MoodleException,
+  MoodleUser,
+  MoodleUserBasic,
+} from '@/app/interfaces/Moodle';
+import { getSearchQuery, getUniqueID } from '../common';
+import {
+  MOODLE_API_URL,
+  MOODLE_HOST_URL,
+  MOODLE_SECRET,
+  MOODLE_STUDENT_ROLE_ID,
+} from './constants';
+import isAuthenticated from '../auth/isAuthenticated';
+import { NextRequest } from 'next/server';
+import AuthErrorResponse from '@/app/interfaces/AuthErrorResponse';
+
+export async function getMoodleUserByEmail(
+  email: string
+): Promise<MoodleUserBasic | null> {
+  try {
+    const url = `${MOODLE_API_URL}?
+        ${getSearchQuery({
+          wstoken: MOODLE_SECRET,
+          wsfunction: 'core_user_get_users_by_field',
+          field: 'email',
+          'values[]': email,
+          moodlewsrestformat: 'json',
+        })}`;
+
+    const res = await fetch(url);
+    const json: MoodleUserBasic[] | MoodleException = await res.json();
+    if ('exception' in json) {
+      throw new Error(json.message);
+    }
+
+    const user: MoodleUserBasic = json[0];
+
+    return user;
+  } catch (_) {
+    throw new Error('Failed to fetch the moodle user.');
+  }
+}
+
+export async function createMoodleUser({
+  email,
+  name,
+}: {
+  email: string;
+  name: string;
+}): Promise<MoodleUser> {
+  try {
+    const splitName = name.split(/\s+/);
+    const firstname = splitName.slice(0, -1).join(' ');
+    const lastname = splitName.slice(-1).join('');
+    const username = email;
+
+    const password = 'TestPass123!' + getUniqueID();
```

Comment:
Our repo is public, makes us more vulnerable having a part of the password as fixed. We can randomise this bit to improve security.